File: app/src/main/java/com/mch/blekot/services/MicroService.java
package com.mch.blekot.services;

import java.util.Stack;
import android.util.Log;
import android.os.IBinder;
import android.app.Service;
import android.content.Intent;
import android.media.MediaRecorder;
import androidx.annotation.Nullable;

import com.mch.blekot.R;

public class MicroService extends Service {

    double mAmplitude, decibels;
    MediaRecorder mRecorder = new MediaRecorder();
    Stack<Double> decibelsHistory = new Stack<>();

    @Override
    public void onCreate() {
        super.onCreate();
        startMicro();
    }

    @Nullable
    @Override
    public IBinder onBind(Intent intent) {
        measureDecibels();
        return null;
    }

    private void startRecorder(){
        String destPath = getApplicationContext().getExternalFilesDir(null).getAbsolutePath();
        Recorder recorder = new Recorder(destPath);
        recorder.startRecord();
    }

    public void startMicro(){
        if (mRecorder == null) {
            mRecorder = new MediaRecorder();
            mRecorder.setAudioSource(MediaRecorder.AudioSource.MIC);
            mRecorder.setOutputFormat(MediaRecorder.OutputFormat.THREE_GPP);
            mRecorder.setAudioEncoder(MediaRecorder.AudioEncoder.AMR_NB);
            mRecorder.setOutputFile("/dev/null");
            try {
                mRecorder.prepare();
            } catch (java.io.IOException | java.lang.SecurityException e) {
                android.util.Log.e("[Micro]", "IOException: " +
                        android.util.Log.getStackTraceString(e));
            }
            try {
                mRecorder.start();
            } catch (java.lang.SecurityException e) {
                android.util.Log.e("[Micro]", "SecurityException: " +
                        android.util.Log.getStackTraceString(e));
            }
        }
    }

    public void measureDecibels() {
        double amplitude = mRecorder.getMaxAmplitude();
        if (amplitude > 0 && amplitude < 1000000) {
            decibels = convertDb(amplitude);
            if (decibels > 10) {
                decibels = convertDb(mRecorder.getMaxAmplitude());
                if (decibelsHistory.size() >= 12) {
                    if (decibelMedia(decibelsHistory) > 10) {
                        Log.i("dec", "turnOffLight: Aqui se apago la luz de sus ojos");
                        decibelsHistory.clear();
                    }
                    Log.i("dec", "pop: " + decibelsHistory.remove(0));
                }
                Log.i("dec", "push: " + decibelsHistory.push(decibels));
            }
        }
    }

    /*----------------------------function for take ambient decibels----------------------------*/
    public double convertDb(double amplitude) {
        /*
         *Los teléfonos pueden alcanzar hasta 90 db + -
         *getMaxAmplitude devuelve un valor entre 0 y 32767 (en la mayoría de los teléfonos). eso
         *significa que si el db máximo es 90, la presión en el micrófono es 0.6325 Pascal. se hace una
         *comparación con el valor anterior de getMaxAmplitude. necesitamos dividir maxAmplitude con
         *(32767/0.6325) 51805.5336 o si son 100db la max entonces 46676.6381
         */
        double EMA_FILTER = 0.6;
        /*SharedPreferences sp = this.getSharedPreferences("device-base", MODE_PRIVATE);
         *double amp = (double) sp.getFloat("amplitude", 0);
         */
        double mEMA = 0.0;
        double mEMAValue = EMA_FILTER * amplitude + (1.0 - EMA_FILTER) * mEMA;
        //Asumiendo que la presión de referencia mínima es 0.000085 Pascal
        // (en la mayoría de los teléfonos) es igual a 0 db
        return 20 * (float) Math.log10((mEMAValue / 51805.5336) / 0.000028251);
    }

    private double decibelMedia(Stack<Double> decHist) {
        Double med = 0D;
        for (int i = 0; i < decHist.size(); i++) {
            med += decHist.get(i);
            Log.i("dec", "media: " + med);
        }
        Log.i("dec", "decibelMedia: " + med / decHist.size());
        return Math.round(((med / decHist.size()) * 100d) / 100d);
    }
}

File: app/src/main/java/com/mch/blekot/services/MicroService.kt
package com.mch.blekot.services

import java.util.*

import android.util.Log
import android.os.Build
import kotlin.math.log10
import android.os.IBinder
import android.app.Service
import kotlinx.coroutines.*
import android.content.Intent
import kotlin.math.roundToLong
import android.content.Context
import android.media.MediaRecorder
import androidx.annotation.RequiresApi


class MicroService : Service() {

    private var mRecorder: MediaRecorder? = null
    private var decibels = 0.0
    private val TAG = "MicroService"
    private var mEMA = 0.0

    private var decibelsHistory: Stack<Double> = Stack()


    override fun onBind(p0: Intent?): IBinder? {
        return null
    }

    @RequiresApi(Build.VERSION_CODES.Q)
    override fun onCreate() {
        Log.d(TAG, "Servicio micro...")
    }


    override fun onStartCommand(intent: Intent?, flags: Int, startId: Int): Int {
        Log.d(TAG, "Servicio iniciado...")


        return START_NOT_STICKY
    }

    /*----------------------------function for take ambient decibels----------------------------*/
    private fun convertDb(amplitude: Double): Double {
        /*
         *Los teléfonos celulares pueden alcanzar hasta 90 db + -
         *getMaxAmplitude devuelve un valor entre 0 y 32767 (en la mayoría de los teléfonos). eso
         *significa que si el db máximo es 90, la presión en el micrófono es 0.6325 Pascal. hace una
         *comparación con el valor anterior de getMaxAmplitude. necesitamos dividir maxAmplitude con
         *(32767/0.6325) 51805.5336 o si 100db entonces 46676.6381
         */
        val EMA_FILTER = 0.6

        /*---------------Decibels measure---------------*/
        //static final private double EMA_FILTER = 0.6;

        mEMA = EMA_FILTER * amplitude + (1.0 - EMA_FILTER) * mEMA
        //Asumiendo que la presión de referencia mínima es 0.000085 Pascal
        // (en la mayoría de los teléfonos) es igual a 0 db
        // TODO: Find out the minimum reference in Redmi
        // return 20 * (float) Math.log10((mEMAValue / 51805.5336) / 0.000028251);
        return (20 * log10(mEMA / 51805.5336 / 0.000028251) * 100).roundToLong() / 100.0
    }

    private fun measureDecibels() {
        val amplitude: Double? = mRecorder?.maxAmplitude?.toDouble()
        //Log.i("ron11", "ron11: " + String.format(Locale.US,"%.2f", amplitude));
        if (amplitude!! > 0 && amplitude < 1000000) {
            decibels = convertDb(amplitude)
            Log.i(TAG, "Decibels: $decibels")

            /*
            if (true) { // decibels > Constants.RUIDO_MIN
                if (decibelsHistory.size >= Constants.DECIBEL_DATA_LENGTH) {
                    //count = 0;
                    if (decibelMedia(decibelsHistory) > Constants.RUIDO_MIN) {
                        Ewelink.turnOffLight()
                        Log.i("dec", "turnOffLight: Aqui se apago la luz de sus ojos")
                        //decibelsHistory.clear();
                    }
                }
                //count++;
                //Log.i("dec", "push-"+count+": " + decibelsHistory.push(decibels));
            }

             */
        }
    }

    private fun decibelMedia(decHist: Stack<Double>): Double {
        var med = 0.00
        val sizeTmp: Int = decHist.size
        while (!decHist.isEmpty()) {
            med += decHist.pop()
        }
        return (med / sizeTmp * 100.0 / 100.0).roundToLong().toDouble()
    }

}